cuda_core/docs/design/buffer-fill-redesign.md

```diff
@@ -0,0 +1,350 @@
+    # Plain Python int -> 1-byte (must be 0-255)
+    width = 1
+elif isinstance(value, bytes):
+    width = len(value)
+    if width not in (1, 2, 4):
+        raise ValueError("bytes must be length 1, 2, or 4")
+else:
+    # NumPy scalar or similar
+    dtype = getattr(value, 'dtype', None)
+    if dtype is None:
+        raise TypeError(f"unsupported value type: {type(value)}")
+    width = dtype.itemsize
+    if width not in (1, 2, 4):
+        raise ValueError(f"unsupported dtype itemsize: {width}")
+```
+
+### Cython Improvements
+
+```cython
+from libc.limits cimport UCHAR_MAX, USHRT_MAX, UINT_MAX
+
+cdef inline int _validate_fill_value(int width, int64_t value) except -1:
+    """Validate value fits in unsigned width. Returns 0 on success."""
+    cdef int64_t max_val
+    if width == 1:
+        max_val = UCHAR_MAX
+    elif width == 2:
+        max_val = USHRT_MAX
+    else:  # width == 4
+        max_val = UINT_MAX
+    
+    if value < 0 or value > max_val:
+        raise ValueError(f"value {value} out of range for {width}-byte unsigned fill")
+    return 0
+```
+
+## Migration Path
+
+Since this is before the v0.5.0 release, we can make breaking changes:
+
+1. Remove the `width` parameter from `Buffer.fill()`
+2. Update signature to `fill(self, value, *, stream)`
+3. Add type dispatch logic
+4. Add `StridedMemoryView.fill()` method
+5. Update tests and documentation
+
+## Non-Contiguous StridedMemoryView Handling
+
+**Problem**: `cuMemsetD*Async` only works on contiguous memory. For non-contiguous (strided) views—where there are gaps between elements due to strides—the memset functions cannot be used directly.
+
+**Example of non-contiguous view**:
+```python
+# A 2D array sliced to take every other row
+arr = np.zeros((100, 100), dtype=np.float32)
+view = arr[::2, :]  # Shape (50, 100), but stride[0] = 800 bytes (2 rows)
+# Memory is NOT contiguous—there are 100-element gaps between rows
+```
+
+**Proposed behavior**: Raise an error for non-contiguous views in the initial implementation:
+
+```python
+def fill(self, value, *, stream: Stream | GraphBuilder):
+    """Fill view elements with value (uses view's dtype).
+    
+    Raises
+    ------
+    ValueError
+        If the view is not contiguous (C or F order).
+    """
+    if not self.layout.is_contiguous_any:
+        raise ValueError(
+            "fill() requires a contiguous view. "
+            "Consider creating a contiguous copy first."
+        )
+    # ... proceed with cuMemsetD*Async
+```
+
+**Future enhancement**: A kernel-based fallback could be added later to support strided fills if there's user demand. This would require launching a simple fill kernel that iterates over elements using the stride information.
+
+## Open Questions
+
+1. **Should plain Python `int` default to 1-byte or 4-byte?**
+   - Proposal: 1-byte (consistent with "Buffer is untyped bytes")
+   - Alternative: Error if int > 255, requiring explicit dtype for larger values
```

Comment:
Python integers have a built-in `to_bytes` function that errors if you try to call it with something larger than 1-byte where it makes sense to limit it to 1-byte in my opinion.